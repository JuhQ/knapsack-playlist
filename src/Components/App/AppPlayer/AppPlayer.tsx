import React from "react";
import { Container } from "semantic-ui-react";

import Queue from "../../../Datastructures/queue/queue";
import EntertainmentSystem from "../../EntertainmentSystem/EntertainmentSystem";

interface Props {
  playlistLength: number;
  generationTried: boolean;
  creating: boolean;
  playlist: Queue;
}

const AppPlayer: React.FC<Props> = ({
  playlistLength,
  generationTried,
  creating,
  playlist,
}: Props) => (
  <Container>
    <h2>The playlist will be generated here.</h2>
    <p>
      Either select randomised sample data with the buttons above, or select
      music to your playlist on the left.
    </p>
    <p>
      In either case, the knapsack algorithm will try to generate a playlist
      which of the highest quality and as close as possible with the given
      length. But be aware, quality will override length in determine the
      playlist content, and therefore the playlist might not be as long as you
      hoped.
    </p>
    {playlistLength === 0 && (
      <>Select dataset size and desired playlist length</>
    )}

    {playlistLength > 0 &&
      generationTried &&
      !creating &&
      !playlist.length() && <>Playlist generated but no songs were found</>}

    {creating && (
      <>
        <h2>Generating playlist.</h2>
        Depending on input size and desired playlist length, this might take a
        while.
      </>
    )}
    {playlistLength > 0 && playlist && playlist.length() > 0 && (
      <EntertainmentSystem list={playlist} length={playlistLength} />
    )}
  </Container>
);

export default AppPlayer;
