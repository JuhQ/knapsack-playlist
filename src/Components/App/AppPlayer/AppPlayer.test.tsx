import React from "react";
import { create } from "react-test-renderer";

import Queue from "../../../Datastructures/queue/queue";
import AppPlayer from "./AppPlayer";

describe("AppPlayer", () => {
  it("should render", () => {
    const playlist = new Queue();
    playlist.merge([
      { id: "1", title: "test AppPlayer", seconds: 60 },
      { id: "2", title: "test2 AppPlayer", seconds: 60 },
    ]);
    const component = create(
      <AppPlayer
        playlistLength={120}
        generationTried
        creating
        playlist={playlist}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render select dataset message", () => {
    const playlist = new Queue();

    const component = create(
      <AppPlayer
        playlistLength={0}
        generationTried
        creating
        playlist={playlist}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render select dataset message even when genration is false", () => {
    const playlist = new Queue();

    const component = create(
      <AppPlayer
        playlistLength={1}
        generationTried={false}
        creating={false}
        playlist={playlist}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render with generation tried false", () => {
    const playlist = new Queue();
    playlist.merge([
      { id: "1", title: "test AppPlayer", seconds: 60 },
      { id: "2", title: "test2 AppPlayer", seconds: 60 },
    ]);
    const component = create(
      <AppPlayer
        playlistLength={120}
        generationTried={false}
        creating
        playlist={playlist}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render with creating false", () => {
    const playlist = new Queue();
    playlist.merge([
      { id: "1", title: "test AppPlayer", seconds: 60 },
      { id: "2", title: "test2 AppPlayer", seconds: 60 },
    ]);
    const component = create(
      <AppPlayer
        playlistLength={120}
        generationTried
        creating={false}
        playlist={playlist}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render with generation tried and creating false", () => {
    const playlist = new Queue();
    playlist.merge([
      { id: "1", title: "test AppPlayer", seconds: 60 },
      { id: "2", title: "test2 AppPlayer", seconds: 60 },
    ]);
    const component = create(
      <AppPlayer
        playlistLength={120}
        generationTried={false}
        creating={false}
        playlist={playlist}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
