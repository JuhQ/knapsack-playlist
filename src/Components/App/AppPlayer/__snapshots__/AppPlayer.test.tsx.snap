// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppPlayer should render 1`] = `
<div
  className="ui container"
>
  Generating playlist. Depending on input size and desired playlist length, this might take a while.
  <div
    className="ui bottom aligned grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <h2>
        test AppPlayer
      </h2>
    </div>
    <div
      className="center aligned five wide computer sixteen wide mobile column"
    >
      <span>
        Playlist duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Tried to generate 
        120
         seconds
      </span>
      <br />
      <strong>
        Great success! Playlist is exactly as long as you wanted!
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ‘Œ
        </span>
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ¤˜
        </span>
        <br />
      </strong>
      <span>
        Queue duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Playlist size: 
        2
         songs
      </span>
      <br />
      <span>
        Queue size: 
        2
         songs
      </span>
    </div>
  </div>
  <div
    className="ui grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <div
        className=""
      >
        <div
          className={null}
          id={null}
        />
      </div>
    </div>
    <div
      className="five wide computer sixteen wide mobile column"
    >
      <div
        className="playlist"
      >
        <div
          className="ui items"
        >
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/1/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test AppPlayer
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/2/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test2 AppPlayer
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppPlayer should render select dataset message 1`] = `
<div
  className="ui container"
>
  Select dataset size and desired playlist length
  Generating playlist. Depending on input size and desired playlist length, this might take a while.
</div>
`;

exports[`AppPlayer should render select dataset message even when genration is false 1`] = `
<div
  className="ui container"
/>
`;

exports[`AppPlayer should render with creating false 1`] = `
<div
  className="ui container"
>
  <div
    className="ui bottom aligned grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <h2>
        test AppPlayer
      </h2>
    </div>
    <div
      className="center aligned five wide computer sixteen wide mobile column"
    >
      <span>
        Playlist duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Tried to generate 
        120
         seconds
      </span>
      <br />
      <strong>
        Great success! Playlist is exactly as long as you wanted!
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ‘Œ
        </span>
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ¤˜
        </span>
        <br />
      </strong>
      <span>
        Queue duration: 
        60
         seconds
      </span>
      <br />
      <span>
        Playlist size: 
        2
         songs
      </span>
      <br />
      <span>
        Queue size: 
        1
         songs
      </span>
    </div>
  </div>
  <div
    className="ui grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <div
        className=""
      >
        <div
          className={null}
          id={null}
        />
      </div>
    </div>
    <div
      className="five wide computer sixteen wide mobile column"
    >
      <div
        className="playlist"
      >
        <div
          className="ui items"
        >
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/2/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test2 AppPlayer
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppPlayer should render with generation tried and creating false 1`] = `
<div
  className="ui container"
>
  <div
    className="ui bottom aligned grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <h2>
        test AppPlayer
      </h2>
    </div>
    <div
      className="center aligned five wide computer sixteen wide mobile column"
    >
      <span>
        Playlist duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Tried to generate 
        120
         seconds
      </span>
      <br />
      <strong>
        Great success! Playlist is exactly as long as you wanted!
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ‘Œ
        </span>
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ¤˜
        </span>
        <br />
      </strong>
      <span>
        Queue duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Playlist size: 
        2
         songs
      </span>
      <br />
      <span>
        Queue size: 
        2
         songs
      </span>
    </div>
  </div>
  <div
    className="ui grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <div
        className=""
      >
        <div
          className={null}
          id={null}
        />
      </div>
    </div>
    <div
      className="five wide computer sixteen wide mobile column"
    >
      <div
        className="playlist"
      >
        <div
          className="ui items"
        >
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/1/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test AppPlayer
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/2/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test2 AppPlayer
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppPlayer should render with generation tried false 1`] = `
<div
  className="ui container"
>
  Generating playlist. Depending on input size and desired playlist length, this might take a while.
  <div
    className="ui bottom aligned grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <h2>
        test AppPlayer
      </h2>
    </div>
    <div
      className="center aligned five wide computer sixteen wide mobile column"
    >
      <span>
        Playlist duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Tried to generate 
        120
         seconds
      </span>
      <br />
      <strong>
        Great success! Playlist is exactly as long as you wanted!
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ‘Œ
        </span>
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ¤˜
        </span>
        <br />
      </strong>
      <span>
        Queue duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Playlist size: 
        2
         songs
      </span>
      <br />
      <span>
        Queue size: 
        2
         songs
      </span>
    </div>
  </div>
  <div
    className="ui grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <div
        className=""
      >
        <div
          className={null}
          id={null}
        />
      </div>
    </div>
    <div
      className="five wide computer sixteen wide mobile column"
    >
      <div
        className="playlist"
      >
        <div
          className="ui items"
        >
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/1/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test AppPlayer
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/2/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test2 AppPlayer
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
